src/persist/src/sqlite.rs

```diff
@@ -0,0 +1,154 @@
+    data blob NOT NULL
+);
+";
+
+/// Implementation of [Consensus] over a sqlite database.
+#[derive(Debug)]
+pub struct SqliteConsensus {
+    conn: Arc<Mutex<Connection>>,
+    shard: String,
+}
+
+impl SqliteConsensus {
+    /// Open a sqlite-backed [Consensus] instance at `path`, for the collection
+    /// named `shard`.
+    pub fn open(path: &Path, shard: String) -> Result<Self, Error> {
+        let mut conn = Connection::open(path)?;
+        let tx = conn.transaction()?;
+        let app_id: i32 = tx.query_row("PRAGMA application_id", params![], |row| row.get(0))?;
+        if app_id == 0 {
+            tx.execute_batch(&format!(
+                "PRAGMA application_id = {APPLICATION_ID};
+                 PRAGMA user_version = 1;"
+            ))?;
+            tx.execute_batch(SCHEMA)?;
+        } else if app_id != APPLICATION_ID {
+            return Err(Error::from(format!("invalid application id: {}", app_id)));
+        }
+        tx.commit()?;
+        Ok(SqliteConsensus {
+            conn: Arc::new(Mutex::new(conn)),
+            shard,
+        })
+    }
+}
+
+#[async_trait]
+impl Consensus for SqliteConsensus {
+    async fn head(&self, _deadline: Instant) -> Result<Option<VersionedData>, Error> {
+        let mut conn = self.conn.lock().await;
+        let tx = conn.transaction()?;
+        let result = tx
+            .prepare(
+                "SELECT sequence_number, data FROM consensus
+                 WHERE shard = $shard",
+            )?
+            .query_row(named_params! {"$shard": self.shard}, |row| {
+                let sequence_number = row.get("sequence_number")?;
+                let data: Vec<_> = row.get("data")?;
+                Ok(VersionedData {
+                    seqno: SeqNo(sequence_number),
+                    data,
+                })
+            })
+            .optional()
+            .map_err(|e| e.into());
+
+        result
+    }
+
+    async fn compare_and_set(
+        &mut self,
+        _deadline: Instant,
+        expected: Option<SeqNo>,
+        new: VersionedData,
+    ) -> Result<Result<(), Option<VersionedData>>, Error> {
+        let mut conn = self.conn.lock().await;
+        let tx = conn.transaction()?;
+        let current = tx
+            .prepare(
+                "SELECT sequence_number, data FROM consensus
+                 WHERE shard = $shard",
+            )?
+            .query_row(named_params! {"$shard": self.shard}, |row| {
+                let sequence_number = row.get("sequence_number")?;
+                let data: Vec<_> = row.get("data")?;
+                Ok(VersionedData {
+                    seqno: SeqNo(sequence_number),
+                    data,
+                })
+            })
+            .optional()?;
+
+        let seqno = current.as_ref().map(|c| c.seqno);
+        if seqno != expected {
+            return Ok(Err(current));
+        }
+
+        if let Some(seqno) = seqno {
+            if new.seqno <= seqno {
```

Comment:
we could move the check (presumably reframed as something like `new.seqno <= expected`) before even talking to sqlite right? if there's some way to make that work, let's do it

Reply:
I don't think thats correct, because the value of `expected` could be arbitrarily incorrect right? There's no way to guarantee that `expected` is actually <= the current sequence number iiuc

Reply:
let me rephrase this another way.

Lets say the current seqno is 7.

if someone tries to do a cas with expected = 5, new = 4

we can give them an Err("expected > new") without looking in the database, but the only way to give them a useful value for the current seqno is to look in the database. I felt like it was important to try to return a useful reading of the current seqno as much as possible, but open to changing my mind